lib/ansible/modules/system/listen_ports_facts.py

```diff
@@ -0,0 +1,353 @@
+                else:
+                    name = ''
+                result = dict(pid=int(pid), address=address, port=int(port), protocol=protocol, name=name)
+                if result not in results:
+                    results.append(result)
+            else:
+                raise EnvironmentError('Could not get process information for the listening ports.')
+    return results
+
+
+def applyWhitelist(portspids, whitelist=None):
+    if whitelist is None:
+        whitelist = list()
+    processes = list()
+    for p in portspids:
+        if int(p['port']) not in whitelist and str(p['port']) not in whitelist:
+            process = dict(pid=p['pid'], address=p['address'], port=p['port'], protocol=p['protocol'], name=p['name'], stime=p['stime'], user=p['user'])
+            if process not in processes:
+                processes.append(process)
+    return processes
+
+
+def main():
+
+    module = AnsibleModule(
+        argument_spec=dict(
+            whitelist_tcp=dict(type='list', default=list()),
+            whitelist_udp=dict(type='list', default=list()),
+        ),
+        supports_check_mode=True,
+    )
+
+    if platform.system() != 'Linux':
+        module.fail_json(msg='This module requires Linux.')
+
+    def getPidSTime(pid):
+        ps_cmd = module.get_bin_path('ps', True)
+        rc, ps_output, stderr = module.run_command([ps_cmd, '-o', 'lstart', '-p', str(pid)])
+        stime = ''
+        if rc == 0:
+            for line in ps_output.splitlines():
+                if 'started' not in line:
+                    stime = line
+        return stime
+
+    def getPidUser(pid):
+        ps_cmd = module.get_bin_path('ps', True)
+        rc, ps_output, stderr = module.run_command([ps_cmd, '-o', 'user', '-p', str(pid)])
+        user = ''
+        if rc == 0:
+            for line in ps_output.splitlines():
+                if line != 'USER':
+                    user = line
+        return user
+
+    result = dict(
+        changed=False,
+        ansible_facts=dict(
+            tcp_listen_violations=list(),
+            udp_listen_violations=list(),
+            tcp_listen=list(),
+            udp_listen=list(),
+        ),
+    )
+
+    try:
+        netstat_cmd = module.get_bin_path('netstat', True)
+
+        # which ports are listening for connections?
+        rc, stdout, stderr = module.run_command([netstat_cmd, '-plunt'])
+        if rc == 0:
+            netstatOut = netStatParse(stdout)
+            for p in netstatOut:
+                p['stime'] = getPidSTime(p['pid'])
+                p['user'] = getPidUser(p['pid'])
+                if p['protocol'] == 'tcp':
+                    result['ansible_facts']['tcp_listen'].append(p)
+                elif p['protocol'] == 'udp':
+                    result['ansible_facts']['udp_listen'].append(p)
+    except KeyError as e:
+        module.fail_json(msg=to_native(e))
+
+    # if a TCP whitelist was supplied, determine which if any pids violate it
```

Comment:
I think we should remove this functionality. Facts modules should be focused solely on gathering data, not validating or checking system state. This particular case is a somewhat gray area because it doesn't actually pass/fail based on that state test. After further discussion with other Core team folks, we feel this should be removed in order to avoid the ambiguity. We can add examples on how to do this with tasks.

Reply:
No problems, I'll remove the whitelist args and related code. How shall we go about providing examples on how to use the facts for doing stuff like comparing to a whitelist? (Is this something that would exist in the comments in the code file?)